backend/python/tapa/floorplan_dse.py

```diff
@@ -0,0 +1,201 @@
+  basic_cmd = get_basic_tapa_command(args_dict, output_dir)
+  basic_cmd += [
+    '--work-dir', work_dir,
+    '--run-tapacc',
+    '--run-hls',
+    '--generate-task-rtl',
+  ]
+  return basic_cmd
+
+
+def get_cmd_to_explore_floorplan(
+    args_dict: Dict[str, Any],
+    work_dir: str,
+    output_dir: str,
+    max_slr_limit: int,
+    min_area_limit: float,
+) -> List[str]:
+  """Command to get one design point. Reuse HLS synthesis and enable_synth_util"""
+  basic_cmd = get_basic_tapa_command(args_dict, output_dir)
+  basic_cmd += [
+    '--work-dir', work_dir,
+    '--generate-task-rtl',
+    '--run-floorplanning',
+    '--generate-top-rtl',
+    '--pack-xo',
+    '--max-slr-width-limit', str(max_slr_limit),
+    '--min-area-limit', str(min_area_limit),
+  ]
+  return basic_cmd
+
+
+def run_floorplan_dse(args):
+  """Self-invoke tapac to generate multiple floorplan outputs"""
+  # remove the file opened in argparse
+  args.floorplan_output.close()
+  os.remove(args.floorplan_output.name)
+
+  args_dict = vars(args)
+  args_dict = {k:v for k, v in args_dict.items() if v is not None}
+
+  top_work_dir = args.work_dir
+  try:
+    os.makedirs(top_work_dir, exist_ok=True)
+  except:
+    _logger.error('Fail to make directory at %s. Check the --work-dir option.', top_work_dir)
+    exit(1)
+
+  _logger.info(f'')
+  _logger.info(f'-----------------------------------------------------------------------------')
+  _logger.info(f'DSE preparation: synthesize each task into RTL')
+  _logger.info(f'-----------------------------------------------------------------------------')
+  _logger.info(f'')
+
+  hls_dir = f'{os.getcwd()}/{top_work_dir}/.csynth/'
+  hls_cmd = get_cmd_to_csynth_tasks(args_dict, hls_dir, tempfile.gettempdir())
+  try:
+    subprocess.run(hls_cmd,
+                    stdout=subprocess.PIPE,
+                    check=True,
+                    universal_newlines=True)
+  except subprocess.CalledProcessError as e:
+    _logger.error('Fail to run command: %s', ' '.join(hls_cmd) )
+    exit(1)
+
+  # start exploring the pareto optimal curve
+  min_slr_limit = args.min_slr_width_limit
+  max_slr_limit = args.max_slr_width_limit
+  min_area_limit = args.min_area_limit
+  id = 1
+
+  while min_slr_limit < max_slr_limit:
+    _logger.info(f'')
+    _logger.info(f'-----------------------------------------------------------------------------')
+    _logger.info(f'Start generating design point {id} with --max-slr-width-limit {max_slr_limit}')
+    _logger.info(f'-----------------------------------------------------------------------------')
+    _logger.info(f'')
+
+    output_dir = f'{top_work_dir}/run-{id}'
+    run_dir = f'{output_dir}/run'
+
+    os.makedirs(run_dir, exist_ok=True)
+
+    # reuse the HLS and RTL synthesis results
+    try:
+      subprocess.run(['rsync', '-az', hls_dir+'/', run_dir],
+                    stdout=subprocess.PIPE,
+                    check=True,
+                    universal_newlines=True)
```

Comment:
We should either avoid introducing another dependency (with `shutil.copytree`), or include `rsync` in the list of dependencies.

Reply:
as of now, rsync is necessary to preserve the timestamp of the files, otherwise enable_synth_util will be run again

Reply:
`shutil.copytree` also preserves the timestamp. The only problem is that `dirs_exist_ok` wasn't added to `shutil.copytree` until Python 3.8. Until Python 3.7 reaches EOL in 2023, we would need to `shutil.rmtree` first. `cp -a` has similar problems. I'm fine with the `rsync` solution, but we must declare it as a dependency [here](https://github.com/UCLA-VAST/tapa/blob/241a8cad1eea8bc0234123dc33b8888b45d60594/CMakeLists.txt#L116) and [here](https://github.com/UCLA-VAST/tapa/blob/241a8cad1eea8bc0234123dc33b8888b45d60594/CMakeLists.txt#L131).